music/src/core/visualizer.ts

```diff
@@ -152,7 +152,9 @@ export class Visualizer {
       this.ctx.fillStyle =
           `rgba(${isActive ? this.config.activeNoteRGB : this.config.noteRGB},
           ${opacity})`;
-      this.ctx.fillRect(x, y, w, this.config.noteHeight);
+      // Round values to the nearest integer to avoid partially filled pixels.
+      this.ctx.fillRect(Math.round(x), Math.round(y), Math.round(w),
+        Math.round(this.config.noteHeight));
```

Comment:
Could we maybe round the noteHeight once, outside of the loop? It’ll be the same through and there’s no point rounding it over and over again.